File: java/datastructure/linkedlist/leetcode/AddTwoNumbers.java
package datastructure.linkedlist.leetcode;

/**
 * @author roseduan
 * 两数相加
 */
public class AddTwoNumbers {

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1);
        ListNode cur = dummy, p = l1, q = l2;
        int count = 0;
        while (p != null || q != null) {
            if (p != null) {
                count += p.val;
                p = p.next;
            }
            if (q != null) {
                count += q.val;
                q = q.next;
            }
            cur.next = new ListNode(count % 10);
            cur = cur.next;
            count /= 10;
        }
        if (count > 0) {
            cur.next = new ListNode(count);
        }
        return dummy.next;
    }

    public class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
}

File: cpp/linkedlist/add_two_numbers.cpp
//
// Created by roseduan on 2021/9/18.
// 两数相加

#include "iostream"
#include "stdio.h"
using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution{
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        auto *dummy = new ListNode();
        auto p = dummy;
        int count = 0;
        while (l1 || l2) {
            if (l1) {
                count += l1->val;
                l1 = l1->next;
            }
            if (l2) {
                count += l2->val;
                l2 = l2->next;
            }
            p->next = new ListNode(count % 10);
            count /= 10;
            p = p->next;
        }
        if (count > 0) {
            p->next = new ListNode(count);
        }
        return dummy->next;
    }
};

int main() {
    auto *node1 = new ListNode(8);
    node1->next = new ListNode(9);
    
    auto *node2 = new ListNode(4);
    node2->next = new ListNode(6);
    
    Solution sol;
    ListNode *res = sol.addTwoNumbers(node1, node2);
    while (res) {
        printf("%d\n", res->val);
        res = res->next;
    }
}
